x-pack/plugins/cases/server/services/so_reference_extractor.ts

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { set } from '@elastic/safer-lodash-set';
+import _ from 'lodash';
+import { SavedObject, SavedObjectReference, SavedObjectsUpdateResponse } from '@kbn/core/server';
+
+interface Field {
+  path: string;
+  type: string;
+  name: string;
+}
+
+/**
+ * This class handles remove fields from an object and moving them into the saved object reference fields. It also
+ * handles going the opposite direction to add fields back into an object by setting them to null or if a reference is
+ * found setting them to the value defined in the reference.
+ *
+ * This population of the field is to avoid having to change the UI to look in the references field of saved objects
+ * to find these values.
+ */
+export class SOReferenceExtractor {
```

Comment:
Same class your class in `x-pack/plugins/security_solution/server/lib/timeline/utils/migrator.ts`. I copy-pasted the code and I renamed the class.